addons/packetfence-perl/install_cpan.sh

```diff
@@ -10,35 +11,98 @@ if [[ ! -f $CsvFile || "$CsvFile" == "" ]]; then
   exit 99
 fi
 
-# ===== PREPARE ENV =====
-mkdir -p /usr/local/pf/lib/perl_modules/lib/perl5/
-export PERL5LIB=/root/perl5/lib/perl5:/usr/local/pf/lib/perl_modules/lib/perl5/
-export PKG_CONFIG_PATH=/usr/lib/pkgconfig/
-TestPerlConfig=$(perl -e exit)
-if [[ "$TestPerlConfig" != "" ]]; then
-  export LC_CTYPE=en_US.UTF-8
-  export LC_ALL=en_US.UTF-8
-fi
+# ===== FUNCTIONS =====
+configure_and_check() {
+    ### Variables
+    # dir of current script
+    SCRIPT_DIR=$(readlink -e $(dirname ${BASH_SOURCE[0]}))
+    BASE_DIR=/usr/local/pf/lib/perl_modules
+    DUMP_FILE=${BASE_DIR}/modules_installed.csv
+    DISABLE_REPO="--disablerepo=packetfence"
+    CPAN_BIN_PATH="/usr/bin/cpan"
+    CPAN_VERSION=2.29
+
+    prepare_env
+    install_requirements
+
+}
+
+prepare_env() {
+    # ===== PREPARE ENV =====
+    mkdir -p ${BASE_DIR}/lib/perl5/
+    # to find already downloaded Perl modules
+    export PERL5LIB=/root/perl5/lib/perl5:${BASE_DIR}/lib/perl5/
+    export PKG_CONFIG_PATH=/usr/lib/pkgconfig/
+    TestPerlConfig=$(perl -e exit)
+    if [[ "$TestPerlConfig" != "" ]]; then
+        export LC_CTYPE=en_US.UTF-8
+        export LC_ALL=en_US.UTF-8
+    fi
+}
+
+install_requirements() {
+    if [ -f /etc/debian_version ]; then
+        echo "Debian system detected"
+        apt update
+        apt install libmodule-signature-perl zip make build-essential libssl-dev zlib1g-dev libmariadb-dev-compat libmariadb-dev libssh2-1-dev libexpat1-dev pkg-config libkrb5-dev libsystemd-dev libgd-dev libcpan-distnameinfo-perl libyaml-perl curl wget graphviz libio-socket-ssl-perl libnet-ssleay-perl libcpan-perl-releases-perl -y
+    elif [ -f /etc/redhat-release ]; then
+        echo "EL system detected"
+        yum install -y openssl-devel krb5-libs MariaDB-devel epel-release libssh2-devel systemd-devel gd-devel perl-open perl-Test perl-experimental perl-CPAN perl-IO-Socket-SSL perl-Net-SSLeay perl-Devel-Peek perl-CPAN-DistnameInfo $DISABLE_REPO
+        yum group install -y "Development Tools" $DISABLE_REPO
+        yum install -y http://repo.okay.com.mx/centos/8/x86_64/release/okay-release-1-5.el8.noarch.rpm
+    else
+        echo "Unknown system, exit"
+        exit 0
+    fi
+}
+
+upgrade_cpan() {
+    echo "CPAN upgrade"
+    # 1. configure CPAN with defaults (answer yes)
+    # 2. override default conf to UNINST cpan after upgrade (seems mandatory for EL8)
+    (echo o conf make_install_arg 'UNINST=1'; echo o conf commit)|PERL_MM_USE_DEFAULT=1 ${CPAN_BIN_PATH} &> /dev/null
+
+    # upgrade CPAN
+    ${CPAN_BIN_PATH} -i ANDK/CPAN-${CPAN_VERSION}.tar.gz &> /dev/null
+
+    # display CPAN version
+    ${CPAN_BIN_PATH} -D CPAN
+    echo "CPAN upgraded"
```

Comment:
Perhaps check the output of the previous line before the echo, no?

Reply:
Following line: https://github.com/inverse-inc/packetfence/blob/0c928bf32d48a03de8c1aa764d8606cd5a7969cc/addons/packetfence-perl/install_cpan.sh#L2

will stop script if `cpan` exit with a return code different from 0. So this line will never be displayed if there is an error.

Reply:
Hum, I wasn't thinking about an error more about if it is an upgrade or not. Like compare the versions, or something like that.

Reply:
I only adjusted text because:
- CPAN installation is idempotent: it will reinstall CPAN at specific version even if CPAN is already installed at that version
- This texte should not appear if there is an error because script will exit before